### ISSUE TITLE
infer enum of sum types

### ISSUE THREAD

**Issue by kahsa** [vlang/v#16022]
**V version: 0.3.1 fa20d79** 
**OS: ArchLinux** 

**What did you do?**
```v
enum En{
	a
	b
	c
}

type EnS = En | string

fn foo(e EnS){
	println(e)
}

fn main(){
	foo('A')
	foo(.a)
	mut arr := []EnS{}
	arr << 'A'
	arr << .a
}
```
**What did you expect to see?**
No error, or this may be a specification.
If so, the error message should be like " sum type of enum cannot omit the enum name."

However, if there are no duplicate enum values in the sum type, it would be very helpful if this could be accepted.
 
**What did you see instead?**
```
./bug.v:15:6: error: expected type is not an enum (`EnS`)
   13 | fn main(){
   14 |     foo('A')
   15 |     foo(.a)
      |         ~~
   16 |     mut arr := []EnS{}
   17 |     arr << 'A'
./bug.v:18:9: error: expected type is not an enum (`EnS`)
   16 |     mut arr := []EnS{}
   17 |     arr << 'A'
   18 |     arr << .a
      |            ~~
   19 | }
```

**Comment by kahsa** [comment]
```v
enum EnFoo{
	a
	b
}

enum EnBar{
	c
	d
}

type EnFB = EnFoo | EnBar

fn foo(e EnFB){
	println(e)
}

fn main(){
	foo(.a)
}
```
This is possible.
But next case is impossible.
```v
enum EnFoo{
	a
	b
}

enum EnBar{
	b
	c
	d
}

type EnFB = EnFoo | EnBar

fn foo(e EnFB){
	println(e)
}

fn main(){
	foo(.b) // must be foo(EnBar.b)
}
```

**Comment by spytheman** [comment]
> However, if there are no duplicate enum values in the sum type, it would be very helpful if this could be accepted.

No, it would not - that is error prone, and can lead to confusion, when the sumtype is later changed.

**Comment by kahsa** [comment]
If so, this error message is a bit confusing.
It should be corrected to a clearer error message.
In this context, `.a` is indeed `EnS` to the human eye.

**Comment by kahsa** [comment]
In the first example, since En is the only candidate, we should be able to determine that .a is En.

In the second actual example, logically, the argument of foo() should be EnFoo or EnBar, so it should be possible to find .b from either Enum and determine that it is the one that exists.

If it is present in both Enums, the error will state that it is not possible to determine which one.

Whether it is hard to implement or not..

**Comment by kahsa** [comment]
> when the sumtype is later changed.

later? at run-time?
I would think that even if the sum types composition is changed later, the incorrect code would still be detected at the time of compilation..

**Comment by kahsa** [comment]
Mistakes detected by the compiler are perfectly acceptable.
What is problematic are errors that are not detected by the compiler and occur at run-time.

**Comment by spytheman** [comment]
In the first example, the problem is that it would require analyzing each place, that uses this sumtype for each enum variant, to see if it has that enum value, and to check if there are ambiguities, when multiple enum values have the same names 
[6:46]spytheman: It will also make code brittle, since once you add a new enum value in any of the enums in the sumtype, that conflicts with some of the existing ones, you would have to go disambiguate all callsites, that use that sumtype and the existing enum value.

> later? at run-time?

No, I mean if you change the source code, as it evolves to support other use cases.

**Comment by spytheman** [comment]
@medvednikov what do you think?

**Comment by spytheman** [comment]
My bad for closing the issue, instead of re-tagging it from a "bug" to "feature request" yesterday.

**Comment by kahsa** [comment]
A feature request might be more appropriate.
At least, if the decision is made, intentionally, that the compiler does not support this syntax, the error message should be changed.

**Comment by kahsa** [comment]
> It will also make code brittle, since once you add a new enum value in any of the enums in the sumtype, that conflicts with some of the existing ones,

Indeed, such weaknesses are quite possible.
But it would be a compilation error, wouldn't it?
I personally think that's enough.

However, I cannot judge whether there is a problem or not in terms of compile speed, compiler complexity, etc.
